### ISSUE TITLE
[cloud_firestore] Snapshots() does not listen to updates from Firestore after the initial connection using a compound query with an equality and inequality simultaneously

### ISSUE THREAD

**Issue by brianbhsu** [FirebaseExtended/flutterfire#2320]
**Describe the bug**
Query.snapshots() does not listen to updates from Firestore after the client initially connects when a compound query using both an equality and an inequality query. On the initial snapshots() call, the local client syncs with Firestore, however on all subsequent changes to the listened documents/collections, the local client does not receive those changes.

**To Reproduce**
1. Create a Stream<QuerySnapshot> by running the snapshots() function on a compound query with both an inequality and an equality query.
```
Firestore.instance.collection('match').where("user_id.$userId",
isGreaterThanOrEqualTo: 0)
        .where("status", isEqualTo: 0)
        .snapshots()
```

**Expected behavior**

The client should be subscribed to Firestore and when a document meets the query criteria (in this case a compound query with an inequality and an equality query), it should be returned.

**Additional context**
```
[✓] Flutter (Channel stable, v1.12.13+hotfix.9, on Mac OS X 10.15.3 19D76, locale en-US)
    • Flutter version 1.12.13+hotfix.9 at /Users/bhsu/home/tools/flutter
    • Framework revision f139b11009 (8 days ago), 2020-03-30 13:57:30 -0700
    • Engine revision af51afceb8
    • Dart version 2.7.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/bhsu/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Community Edition (version 2018.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 31.3.2
    • Dart plugin version 182.4129.13

[✓] VS Code (version 1.43.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.0
```

**Comment by TahaTesser** [comment]
Hi @brianbhsu 
can you please provide your `flutter run --verbose'?
Thank you

**Comment by brianbhsu** [comment]
@TahaTesser Which part, the initial build, when the bug occurs, or everything?

**Comment by Ehesp** [comment]
I've been taking a quick look into this one, and I can't understand this section: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/method_channel/method_channel_query.dart#L64

It seems to be waiting for the Future twice - but in the second one it assigning the query observer handler ID. Looks like there could be a race condition & bug here.

**Comment by alvinthen** [comment]
I was facing the same issue until I run `flutter run --verbose`, there's a failure message regarding the query requires an index and its URL. After creating it, I am now able to listen for changes

**Comment by Salakar** [comment]
Hey all, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
